lib/auth/default_current_user_provider.rb

```diff
@@ -128,33 +141,43 @@ def current_user
     # possible we have an api call, impersonate
     if api_key
       current_user = lookup_api_user(api_key, request)
-      raise Discourse::InvalidAccess.new(I18n.t('invalid_api_credentials'), nil, custom_message: "invalid_api_credentials") unless current_user
+      if !current_user
+        raise Discourse::InvalidAccess.new(
+          I18n.t('invalid_api_credentials'),
+          nil,
+          custom_message: "invalid_api_credentials"
+        )
+      end
       raise Discourse::InvalidAccess if current_user.suspended? || !current_user.active
+      admin_api_key_limiter.performed! if !Rails.env.profile?
       @env[API_KEY_ENV] = true
-      rate_limit_admin_api_requests!
     end
 
     # user api key handling
     if user_api_key
-
       hashed_user_api_key = ApiKey.hash_key(user_api_key)
-      limiter_min = RateLimiter.new(nil, "user_api_min_#{hashed_user_api_key}", GlobalSetting.max_user_api_reqs_per_minute, 60)
-      limiter_day = RateLimiter.new(nil, "user_api_day_#{hashed_user_api_key}", GlobalSetting.max_user_api_reqs_per_day, 86400)
+      @env[HASHED_USER_API_KEY] = hashed_user_api_key
 
-      unless limiter_day.can_perform?
-        limiter_day.performed!
-      end
+      user_api_key_obj = UserApiKey
+        .active
+        .with_key(user_api_key)
+        .includes(:user, :scopes)
+        .first
 
-      unless  limiter_min.can_perform?
-        limiter_min.performed!
-      end
+      raise Discourse::InvalidAccess unless user_api_key_obj
+
+      user_api_key_limiter_60_secs.performed!
+      user_api_key_limiter_1_day.performed!
+
+      user_api_key_obj.ensure_allowed!(@env)
 
-      current_user = lookup_user_api_user_and_update_key(user_api_key, @env[USER_API_CLIENT_ID])
-      raise Discourse::InvalidAccess unless current_user
+      current_user = user_api_key_obj.user
+      raise Discourse::InvalidAccess if !current_user
```

Comment:
I think we can skip this here if we do an inner join on the users table when fetching the `UserApiKey`